core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/FlexibleChecksummer.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.internal.signer;
+
+import static software.amazon.awssdk.http.auth.aws.internal.util.ChecksumUtil.fromChecksumAlgorithm;
+import static software.amazon.awssdk.http.auth.aws.internal.util.ChecksumUtil.readAll;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.getBinaryRequestPayloadStream;
+
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.checksums.spi.ChecksumAlgorithm;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.auth.aws.internal.checksums.SdkChecksum;
+import software.amazon.awssdk.http.auth.aws.internal.io.ChecksumInputStream;
+import software.amazon.awssdk.http.auth.aws.internal.io.ChecksumSubscriber;
+import software.amazon.awssdk.http.auth.aws.signer.Checksummer;
+
+/**
+ * A "flexible" implementation of a checksummer. It takes a map of checksums and their header names, computes them efficiently
+ * by updating each checksum while reading the payload (once), and adds the computed checksum strings to the request using the
+ * given header names in the map. This should be used in cases where a (flexible) checksum algorithm is present during signing.
+ */
+@SdkInternalApi
+public final class FlexibleChecksummer implements Checksummer {
+    private final Map<String, SdkChecksum> checksumMap;
+
+    public FlexibleChecksummer(Map<String, ChecksumAlgorithm> algorithmMap) {
+        this.checksumMap = algorithmMap
+            .entrySet()
+            .stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, v -> fromChecksumAlgorithm(v.getValue())));
+    }
+
+    @Override
+    public void checksum(ContentStreamProvider payload, SdkHttpRequest.Builder request) {
```

Comment:
Instead of passing request builder and letting each implementation decide what to do with it, can we pass how we want to consume the string to it, i.e, `Consumer<String> checksumConsumer`? This way, `X_AMZ_CONTENT_SHA256` is not leaked in checksum implementations and it gives us flexibility to support a new header or new field.

Something like this:

```java
    public FlexibleChecksummer(Map<Consumer<String>, ChecksumAlgorithm> algorithmMap) {
        this.checksumMap = algorithmMap
            .entrySet()
            .stream()
            .collect(Collectors.toMap(Map.Entry::getKey, v -> fromChecksumAlgorithm(v.getValue())));
    }

    @Override
    public void checksum(ContentStreamProvider payload) {
        InputStream payloadStream = getBinaryRequestPayloadStream(payload);

        ChecksumInputStream computingStream = new ChecksumInputStream(
            payloadStream,
            checksumMap.values()
        );

        readAll(computingStream);

        checksumMap.forEach((consumer, checksum) -> consumer.accept(checksum.getChecksum()));
    }

```

Reply:
I'm not suggesting passing request instance to this class during initialization though; I'm suggesting passing how we consume the request builder to this class. 
```
        FlexibleChecksummer checksummer = new FlexibleChecksummer(
            ImmutableMap.of(SHA256, value -> request.putHeader("sha256", value)));
```

IMO, the sole responsibility of checksummer class should be calculating checksum and now it seems it does multiple things and violates SRP, which makes it harder to extend in the future. For example, we may need to save the checksum in a different header for certain operations. It seems with current implementations, we'd need another checksummer class.

Reply:
Discussed offline, since this is an internal API, having another layer of handling the checksum values seems a bit overkill.